From 7ec135739228d6b18792eaefdb702676e09b6138 Mon Sep 17 00:00:00 2001
From: Lu Feifei <lufeifei@wxiat.com>
Date: Tue, 7 Jun 2022 17:42:02 +0800
Subject: [PATCH] sw64: fix printk method for guest os
Patch-mainline: NO, OTHERS
References: openEuler-22.03-LTS
Modified-by-SEL: No


Sunway inclusion
category: feature
bugzilla: https://gitee.com/openeuler/kernel/issues/I56WV8

--------------------------------

Output kernel log through the specified IO address is required by
qemu, and this method is not required by guest.

Signed-off-by: Lu Feifei <lufeifei@wxiat.com>

Signed-off-by: Gu Zitao <guzitao@wxiat.com>
Acked-by: Xie XiuQi <xiexiuqi@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 arch/sw_64/include/asm/hw_init.h | 1 +
 arch/sw_64/kernel/dup_print.c    | 2 +-
 2 files changed, 2 insertions(+), 1 deletion(-)

diff --git a/arch/sw_64/include/asm/hw_init.h b/arch/sw_64/include/asm/hw_init.h
index ede8da6b0314..594704756e72 100644
--- a/arch/sw_64/include/asm/hw_init.h
+++ b/arch/sw_64/include/asm/hw_init.h
@@ -106,6 +106,7 @@ DECLARE_STATIC_KEY_FALSE(run_mode_emul_key);
 
 #define is_in_host()		static_branch_likely(&run_mode_host_key)
 #define is_in_guest()		static_branch_unlikely(&run_mode_guest_key)
+#define is_in_emul()		static_branch_unlikely(&run_mode_emul_key)
 #define is_guest_or_emul()	!static_branch_likely(&run_mode_host_key)
 
 #define CPU_SW3231		0x31
diff --git a/arch/sw_64/kernel/dup_print.c b/arch/sw_64/kernel/dup_print.c
index fae543625f62..e1c05cfa864b 100644
--- a/arch/sw_64/kernel/dup_print.c
+++ b/arch/sw_64/kernel/dup_print.c
@@ -40,7 +40,7 @@ int sw64_printk(const char *fmt, va_list args)
 		printed_len += vscnprintf(sw64_printk_buf, 1024, fmt, args);
 	} else {
 		printed_len += vscnprintf(sw64_printk_buf, 1024, fmt, args);
-		if (is_guest_or_emul()) {
+		if (is_in_emul()) {
 			unsigned long write_addr = QEMU_PRINTF_BUFF_BASE;
 			*(unsigned long *)write_addr = (unsigned long)((((unsigned long)sw64_printk_buf) & 0xffffffffUL)
 					| ((unsigned long)printed_len << 32));
-- 
2.34.1

